File: src/core/encoder.js
/**
 * encoder.js
 * High-level TX encoder: message -> frames -> symbol grids
 * Uses packet and modulation modules.
 */

import { gridGeom, bytesToFrames, buildFramePayload } from './packet.js';
import { valueToColor, setModulationMode, getCurrentMode } from './modulation.js';

export class Encoder {
  constructor(density = 'medium', canvasW = 800, canvasH = 600) {
    this.density = density;
    this.canvasW = canvasW;
    this.canvasH = canvasH;
    this.margin = 60;
    this.geom = gridGeom(density, canvasW, canvasH, this.margin);
    this.mode = getCurrentMode();
  }

  setDensity(density) {
    this.density = density;
    this.geom = gridGeom(density, this.canvasW, this.canvasH, this.margin);
  }

  setModulation(mode) {
    setModulationMode(mode);
    this.mode = mode;
  }

  /**
   * Encode full message into sequence of frames (each frame is array of symbol values)
   */
  encodeMessage(text) {
    const msgBytes = new TextEncoder().encode(text || '');
    const lenBytes = new Uint8Array(4);
    new DataView(lenBytes.buffer).setUint32(0, msgBytes.length, false);

    const fullBytes = new Uint8Array(lenBytes.length + msgBytes.length);
    fullBytes.set(lenBytes, 0);
    fullBytes.set(msgBytes, 4);

    const { frames: byteFrames, bytesPerFrame, payloadCells } = bytesToFrames(fullBytes, this.density);
    const totalFrames = byteFrames.length;

    const symbolFrames = byteFrames.map((chunk, idx) => {
      return buildFramePayload(idx, totalFrames, chunk, payloadCells);
    });

    return {
      frames: symbolFrames,
      bytesPerFrame,
      payloadCells,
      totalFrames,
      geom: this.geom,
      density: this.density,
      mode: this.mode
    };
  }

  /**
   * Draw a single frame onto a canvas context (for TX)
   * Returns symbol values used.
   */
  /**
   * Optimized draw stub — measurement framework tracks time here
   */
  drawFrameFast(ctx, symbolFrame, geom = this.geom) {
    const t0 = performance.now();
    const result = this.drawFrame(ctx, symbolFrame, geom);
    const t1 = performance.now();
    if (window.photonProfiler) window.photonProfiler.recordFrame({ renderMs: t1 - t0 });
    return result;
  }

  drawFrame(ctx, symbolFrame, geom = this.geom) {
    // Optimized path: build ImageData buffer once instead of thousands of fillRect
    const cellW = Math.ceil(geom.cellW);
    const cellH = Math.ceil(geom.cellH);
    const imgW = geom.w * cellW + 1;
    const imgH = geom.h * cellH + 1;
    const imgData = ctx.createImageData(imgW, imgH);
    const data = imgData.data;

    let vi = 0;
    for (let row = 0; row < geom.h; row++) {
      for (let col = 0; col < geom.w; col++) {
        const v = symbolFrame[vi++];
        const [r, g, b] = valueToColor(v);
        // Fill cell region directly in pixel buffer (simplified)
        for (let dy = 0; dy < cellH; dy++) {
          for (let dx = 0; dx < cellW; dx++) {
            const px = (row * cellH + dy) * imgW * 4 + (col * cellW + dx) * 4;
            if (px + 3 < data.length) {
              data[px] = r; data[px + 1] = g; data[px + 2] = b; data[px + 3] = 255;
            }
          }
        }
      }
    }
    ctx.putImageData(imgData, geom.marginX || 60, geom.marginY || 60);

    // Corner markers
    ctx.fillStyle = 'rgb(255,0,255)';
    const ms = 40;
    const positions = [
      { x: 30, y: 30 }, { x: this.canvasW - 30, y: 30 },
      { x: this.canvasW - 30, y: this.canvasH - 30 }, { x: 30, y: this.canvasH - 30 }
    ];
    for (const p of positions) {
      ctx.fillRect(p.x - ms / 2, p.y - ms / 2, ms, ms);
    }
    return symbolFrame;
  }
}

File: native/rust/src/encoder.rs
//! High-level encoder

use crate::modulation::{ModulationMode, value_to_color};
use crate::packet::{GridGeom, build_frame_payload, bytes_to_frames, HEADER_CELLS};

pub struct Encoder {
    pub width: usize,
    pub height: usize,
    pub margin: f32,
    pub geom: GridGeom,
    pub mode: ModulationMode,
}

impl Encoder {
    pub fn new(grid_w: usize, grid_h: usize, mode: ModulationMode) -> Self {
        let geom = GridGeom::new((grid_w, grid_h), 800.0, 600.0, 60.0);
        Self {
            width: grid_w,
            height: grid_h,
            margin: 60.0,
            geom,
            mode,
        }
    }

    pub fn set_mode(&mut self, mode: ModulationMode) {
        self.mode = mode;
    }

    /// Encode message into list of symbol frames
    pub fn encode_message(&self, text: &str) -> (Vec<Vec<u16>>, usize, usize) {
        let msg_bytes = text.as_bytes();
        let mut len_bytes = [0u8; 4];
        let len = msg_bytes.len() as u32;
        len_bytes.copy_from_slice(&len.to_be_bytes());

        let mut full: Vec<u8> = Vec::with_capacity(4 + msg_bytes.len());
        full.extend_from_slice(&len_bytes);
        full.extend_from_slice(msg_bytes);

        let total_cells = self.width * self.height;
        let payload_cells = total_cells - HEADER_CELLS;

        let byte_frames = bytes_to_frames(&full, payload_cells, self.mode);
        let total_frames = byte_frames.len();

        let mut symbol_frames = Vec::new();
        for (idx, chunk) in byte_frames.iter().enumerate() {
            let frame = build_frame_payload(idx, total_frames, chunk, payload_cells, self.mode);
            symbol_frames.push(frame);
        }

        (symbol_frames, total_frames, payload_cells)
    }

    /// Render a frame to a raw RGB buffer (for native rendering)
    pub fn render_frame(&self, symbols: &[u16], out_width: usize, out_height: usize) -> Vec<u8> {
        let mut buf = vec![0u8; out_width * out_height * 3];
        let cell_w = self.geom.cell_w;
        let cell_h = self.geom.cell_h;

        for row in 0..self.height {
            for col in 0..self.width {
                let idx = row * self.width + col;
                let v = symbols.get(idx).copied().unwrap_or(0);
                let (r, g, b) = value_to_color(v, self.mode);

                let x0 = (self.margin + col as f32 * cell_w) as usize;
                let y0 = (self.margin + row as f32 * cell_h) as usize;
                let w = cell_w.ceil() as usize + 1;
                let h = cell_h.ceil() as usize + 1;

                for y in y0..(y0 + h).min(out_height) {
                    for x in x0..(x0 + w).min(out_width) {
                        let off = (y * out_width + x) * 3;
                        if off + 2 < buf.len() {
                            buf[off] = r;
                            buf[off + 1] = g;
                            buf[off + 2] = b;
                        }
                    }
                }
            }
        }

        // Draw magenta markers
        let ms = 40usize;
        let positions = [
            (30, 30),
            (out_width - 30 - ms, 30),
            (out_width - 30 - ms, out_height - 30 - ms),
            (30, out_height - 30 - ms),
        ];

        for (px, py) in positions {
            for y in py..(py + ms).min(out_height) {
                for x in px..(px + ms).min(out_width) {
                    let off = (y * out_width + x) * 3;
                    if off + 2 < buf.len() {
                        buf[off] = 255;
                        buf[off + 1] = 0;
                        buf[off + 2] = 255;
                    }
                }
            }
        }

        buf
    }
}
